init.lua

```diff
@@ -89,19 +89,51 @@ function torch.type(obj)
 end
 
 --[[ See if a given object is an instance of the provided torch class. ]]
+local function exactTypeMatch(obj_mt, typeSpecNameString)
+  return obj_mt.__typename == typeSpecNameString
+end
+
+local function partialTypeMatch(obj_mt, typeSpecNameString)
```

Comment:
My point was, you don't need the tables for that. All you seem to be checking is that typeSpecNameString is a suffix of obj_mt.__typename AND that this suffix is either preceded by a point '.' or by nothing. Roughly along these lines:
local lenSuffix, lenTotal = #typeSpecNameString, #obj_mt.__typename
if lenSuffix > lenTotal or 
  string.sub(obj_mt.__typename, lenTotal-lenSuffix+1) ~= typeSpecNameString then
  return false
end
if lenTotal == lenSuffix or
  obj_mt.__typename[lenTotal-lenSuffix] == '.' then
  return true
else
  return false
end

That could probably still be streamlined. Perhaps I'm still missing some subtlety but isn't that doing the same thing without any table allocations (and in fact without string reversal, too) ?